File: ruby/day17_part2_2020.rb

require 'set'

ACTIVE = '#'
INACTIVE = '.'

def count_active_neighbors(hypercube, x, y, z, w)
  count = 0
  (-1..1).each do |dw|
    (-1..1).each do |dz|
      (-1..1).each do |dy|
        (-1..1).each do |dx|
          next if dx == 0 && dy == 0 && dz == 0 && dw == 0
          count += 1 if hypercube.include?([x + dx, y + dy, z + dz, w + dw])
        end
      end
    end
  end
  count
end

def simulate_cycle(hypercube)
  new_hypercube = Set.new
  x_range = (hypercube.map { |x, _, _, _| x }.min - 1..hypercube.map { |x, _, _, _| x }.max + 1)
  y_range = (hypercube.map { |_, y, _, _| y }.min - 1..hypercube.map { |_, y, _, _| y }.max + 1)
  z_range = (hypercube.map { |_, _, z, _| z }.min - 1..hypercube.map { |_, _, z, _| z }.max + 1)
  w_range = (hypercube.map { |_, _, _, w| w }.min - 1..hypercube.map { |_, _, _, w| w }.max + 1)

  x_range.each do |x|
    y_range.each do |y|
      z_range.each do |z|
        w_range.each do |w|
          active_neighbors = count_active_neighbors(hypercube, x, y, z, w)
          if hypercube.include?([x, y, z, w])
            new_hypercube << [x, y, z, w] if active_neighbors.between?(2, 3)
          else
            new_hypercube << [x, y, z, w] if active_neighbors == 3
          end
        end
      end
    end
  end

  new_hypercube
end

hypercube = Set.new
File.open('input.txt').each_with_index do |line, y|
  line.strip.chars.each_with_index do |char, x|
    hypercube << [x, y, 0, 0] if char == ACTIVE
  end
end

6.times do
  hypercube = simulate_cycle(hypercube)
end

puts hypercube.size

File: typescript/day17_part2_2020.ts
import * as fs from 'fs';

type CubeState = 'active' | 'inactive';

class PocketDimension {
    private cubes: Map<string, CubeState> = new Map();
    private directions: number[][];

    constructor(initialState: string[]) {
        this.directions = this.generateDirections();
        this.initializeCubes(initialState);
    }

    private generateDirections(): number[][] {
        const directions = [];
        for (let x = -1; x <= 1; x++) {
            for (let y = -1; y <= 1; y++) {
                for (let z = -1; z <= 1; z++) {
                    for (let w = -1; w <= 1; w++) {
                        if (x !== 0 || y !== 0 || z !== 0 || w !== 0) {
                            directions.push([x, y, z, w]);
                        }
                    }
                }
            }
        }
        return directions;
    }

    private initializeCubes(initialState: string[]) {
        initialState.forEach((line, y) => {
            for (let x = 0; x < line.length; x++) {
                const state = line[x] === '#' ? 'active' : 'inactive';
                this.cubes.set(`${x},${y},0,0`, state);
            }
        });
    }

    private getActiveNeighbors(x: number, y: number, z: number, w: number): number {
        return this.directions.reduce((count, [dx, dy, dz, dw]) => {
            const neighborKey = `${x + dx},${y + dy},${z + dz},${w + dw}`;
            return count + (this.cubes.get(neighborKey) === 'active' ? 1 : 0);
        }, 0);
    }

    public simulateCycles(cycles: number) {
        for (let cycle = 0; cycle < cycles; cycle++) {
            const newCubes = new Map<string, CubeState>();
            const keys = Array.from(this.cubes.keys());
            const bounds = this.calculateBounds(keys);

            for (let x = bounds.minX - 1; x <= bounds.maxX + 1; x++) {
                for (let y = bounds.minY - 1; y <= bounds.maxY + 1; y++) {
                    for (let z = bounds.minZ - 1; z <= bounds.maxZ + 1; z++) {
                        for (let w = bounds.minW - 1; w <= bounds.maxW + 1; w++) {
                            const activeNeighbors = this.getActiveNeighbors(x, y, z, w);
                            const currentKey = `${x},${y},${z},${w}`;
                            const currentState = this.cubes.get(currentKey) || 'inactive';

                            if (currentState === 'active' && (activeNeighbors === 2 || activeNeighbors === 3)) {
                                newCubes.set(currentKey, 'active');
                            } else if (currentState === 'inactive' && activeNeighbors === 3) {
                                newCubes.set(currentKey, 'active');
                            } else {
                                newCubes.set(currentKey, 'inactive');
                            }
                        }
                    }
                }
            }
            this.cubes = newCubes;
        }
    }

    private calculateBounds(keys: string[]) {
        const coordinates = keys.map(key => key.split(',').map(Number));
        const xs = coordinates.map(coord => coord[0]);
        const ys = coordinates.map(coord => coord[1]);
        const zs = coordinates.map(coord => coord[2]);
        const ws = coordinates.map(coord => coord[3]);

        return {
            minX: Math.min(...xs),
            maxX: Math.max(...xs),
            minY: Math.min(...ys),
            maxY: Math.max(...ys),
            minZ: Math.min(...zs),
            maxZ: Math.max(...zs),
            minW: Math.min(...ws),
            maxW: Math.max(...ws),
        };
    }

    public countActiveCubes(): number {
        return Array.from(this.cubes.values()).filter(state => state === 'active').length;
    }
}

const main = async () => {
    const input = fs.readFileSync('input.txt', 'utf-8').trim().split('\n');
    const pocketDimension = new PocketDimension(input);
    pocketDimension.simulateCycles(6);
    console.log(pocketDimension.countActiveCubes());
};

main();